bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -513,13 +509,15 @@ void ReplicaImp::onMessage<ClientRequestMsg>(std::unique_ptr<ClientRequestMsg> m
           }
           requestsOfNonPrimary[msg->requestSeqNum()] = std::make_pair(getMonotonicTime(), msg);
         }
-        send(msg, currentPrimary());
+
+        send(m.release(), currentPrimary());
```

Comment:
Don't we have a leak here?
I didnt see any release inside the chain of the above calls.
So you are releasing and leaking here.

Reply:
The memory management is the same in that case, so it there was no leaks, it would cause one.

Reply:
One thing is dangerous here. The send is asynchronous, so it might occur that message is deleted before it was send.

Reply:
@titusen  the send is async, but the data is being copy to a vector.
message can be deleted any time after returning.
You don't need to worry, the pointer can be deleted after call returns.
See:
```
int MsgsCommunicator::sendAsyncMessage(NodeNum destNode, char* message, size_t messageLength) {
  return communication_->send(destNode, std::vector<uint8_t>(message, message + messageLength));
}
```

Reply:
Done